File: WowUp.WPF/AddonProviders/Curse/CurseFolderScanner.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using WowUp.Common.Extensions;

namespace WowUp.WPF.AddonProviders.Curse
{
    /// <summary>
    /// Parse a single addon directory
    /// </summary>
    public class CurseFolderScanner
    {
        private readonly DirectoryInfo _addonDirectory;

        private Regex TocFileCommentsRegex() => new Regex(@"(?m)\s*#.*$");
        private Regex TocFileIncludesRegex() => new Regex(@"(?mi)^\s*((?:(?<!\.\.).)+\.(?:xml|lua))\s*$");
        private Regex TocFileRegex() => new Regex(@"(?i)^([^\/]+)[\\\/]\1\.toc$");

        private Regex BindingsXmlCommentsRegex() => new Regex(@"(?s)<!--.*?-->");
        private Regex BindingsXmlIncludesRegex() => new Regex(@"(?i)<(?:Include|Script)\s+file=[\""\""']((?:(?<!\.\.).)+)[\""\""']\s*/>");
        private Regex BindingsXmlRegex() => new Regex(@"(?i)^[^\/\\]+[\/\\]Bindings\.xml$");

        public int FileCount { get; private set; }
        public long FileDateHash { get; private set; }
        public long Fingerprint { get; private set; }
        public string FolderName { get; private set; }
        public List<long> IndividualFingerprints { get; private set; }
        public DateTime LastWriteTimeUtc { get { return _addonDirectory.LastWriteTimeUtc; } }

        public CurseFolderScanner(DirectoryInfo addonDirectory)
        {
            _addonDirectory = addonDirectory;
        }

        private static bool FilePathHasInvalidChars(string path)
        {
            return !string.IsNullOrEmpty(path) && path.IndexOfAny(Path.GetInvalidPathChars()) >= 0;
        }

        public async Task<CurseFolderScanner> ScanFolder()
        {
            var fileSystemInfos = _addonDirectory.GetFileSystemInfos();
            FileCount = fileSystemInfos.Length;
            FileDateHash = CurseHash.GetFileDateHash(fileSystemInfos);

            var matchingFiles = await GetMatchingFiles(_addonDirectory);
            matchingFiles.Sort();

            IndividualFingerprints = new List<long>();
            //var tuples = new List<(long, string)>();
            foreach (string path in matchingFiles)
            {
                var normalizedFileHash = CurseHash.ComputeNormalizedFileHash(path);
                IndividualFingerprints.Add(normalizedFileHash);
                //tuples.Add((normalizedFileHash, path));
            }

            IndividualFingerprints.Sort();

            //tuples.Sort((a, b) => a.Item1.CompareTo(b.Item1));

            var hashConcat = string.Join(string.Empty, IndividualFingerprints);
            Fingerprint = CurseHash.ComputeHash(Encoding.ASCII.GetBytes(hashConcat), false);

            return this;
        }

        private async Task<List<string>> GetMatchingFiles(DirectoryInfo directory)
        {
            var matchingFileList = new List<string>();
            var fileInfoList = new List<FileInfo>();

            string str = _addonDirectory.Parent.FullName + Path.DirectorySeparatorChar.ToString();
            foreach (FileInfo file in directory.GetFiles("*.*", SearchOption.AllDirectories))
            {
                string input = file.FullName.ToLower().Replace(str.ToLower(), "");
                if (TocFileRegex().Match(input).Success)
                {
                    fileInfoList.Add(file);
                }

                if (BindingsXmlRegex().Match(input).Success)
                {
                    matchingFileList.Add(file.FullName.ToLowerInvariant());
                }
            }

            foreach (var fileInfo in fileInfoList)
            {
                await ProcessIncludeFile(matchingFileList, fileInfo);
            }

            return matchingFileList;
        }

        private async Task ProcessIncludeFile(List<string> matchingFileList, FileInfo fileInfo)
        {
            if (!fileInfo.Exists || matchingFileList.Contains(fileInfo.FullName.ToLowerInvariant()))
            {
                return;
            }

            matchingFileList.Add(fileInfo.FullName.ToLowerInvariant());

            string input = await fileInfo.ReadTextAsync();
            //using (var streamReader = new StreamReader(fileInfo.FullName))
            //{
            //    input = streamReader.ReadToEnd();
            //    streamReader.Close();
            //}

            input = RemoveComments(fileInfo, input);

            var inclusions = GetFileInclusionMatches(fileInfo, input);
            if (inclusions == null)
            {
                return;
            }

            foreach (Match match in inclusions)
            {
                string fileName;
                string str = match.Groups[1].Value;
                if (FilePathHasInvalidChars(str))
                {
                    // ERROR
                }

                fileName = Path.Combine(fileInfo.DirectoryName, str);

                await ProcessIncludeFile(matchingFileList, new FileInfo(fileName));
            }
        }

        private MatchCollection GetFileInclusionMatches(FileInfo fileInfo, string fileContent)
        {
            return fileInfo.Extension switch
            {
                ".xml" => BindingsXmlIncludesRegex().Matches(fileContent),
                ".toc" => TocFileIncludesRegex().Matches(fileContent),
                _ => null,
            };
        }

        private string RemoveComments(FileInfo fileInfo, string fileContent)
        {
            return fileInfo.Extension switch
            {
                ".xml" => BindingsXmlCommentsRegex().Replace(fileContent, string.Empty),
                ".toc" => TocFileCommentsRegex().Replace(fileContent, string.Empty),
                _ => fileContent,
            };
        }
    }
}

File: wowup-electron/src/app/addon-providers/curse/curse-folder-scanner.ts
import { FileService } from "app/services/files/file.service";
import * as path from 'path';
import * as fs from 'fs';
import * as _ from 'lodash';
import { AddonFolder } from "app/models/wowup/addon-folder";
import { ElectronService } from "app/services";
import { CurseHashFileResponse } from "common/models/curse-hash-file-response";
import { CurseHashFileRequest } from "common/models/curse-hash-file-request";
import { CURSE_HASH_FILE_CHANNEL } from "common/constants";
import { v4 as uuidv4 } from 'uuid';
import { CurseScanResult } from "../../models/curse/curse-scan-result";
import { from } from "rxjs";
import { mergeMap } from "rxjs/operators";

export class CurseFolderScanner {

    constructor(
        private _electronService: ElectronService,
        private _fileService: FileService
    ) { }

    private get tocFileCommentsRegex() {
        return /\s*#.*$/mg;
    }

    private get tocFileIncludesRegex() {
        return /^\s*((?:(?<!\.\.).)+\.(?:xml|lua))\s*$/mig;
    }

    private get tocFileRegex() {
        return /^([^\/]+)[\\\/]\1\.toc$/i;
    }

    private get bindingsXmlRegex() {
        return /^[^\/\\]+[\/\\]Bindings\.xml$/i;
    }

    private get bindingsXmlIncludesRegex() {
        return /<(?:Include|Script)\s+file=[\""\""']((?:(?<!\.\.).)+)[\""\""']\s*\/>/ig;
    }

    private get bindingsXmlCommentsRegex() {
        return /<!--.*?-->/gs;
    }

    async scanFolder(addonFolder: AddonFolder): Promise<CurseScanResult> {
        const folderPath = addonFolder.path;
        const files = await this._fileService.listAllFiles(folderPath);
        console.log('listAllFiles', folderPath, files.length);

        let matchingFiles = await this.getMatchingFiles(folderPath, files);
        matchingFiles = _.sortBy(matchingFiles, f => f.toLowerCase());

        console.log('matching files', matchingFiles.length)
        // const fst = matchingFiles.map(f => f.toLowerCase()).join('\n');

        const individualFingerprints: number[] = [];
        for (let path of matchingFiles) {
            const normalizedFileHash = await this.computeNormalizedFileHash(path);
            individualFingerprints.push(normalizedFileHash);
        }

        const hashConcat = _.orderBy(individualFingerprints).join('');
        const fingerprint = await this.computeStringHash(hashConcat);
        console.log('fingerprint', fingerprint);

        return {
            directory: folderPath,
            fileCount: matchingFiles.length,
            fingerprint,
            folderName: path.basename(folderPath),
            individualFingerprints,
            addonFolder
        };
    }

    private async getMatchingFiles(folderPath: string, filePaths: string[]): Promise<string[]> {
        const parentDir = path.dirname(folderPath) + path.sep;
        const matchingFileList: string[] = [];
        const fileInfoList: string[] = [];
        for (let filePath of filePaths) {
            const input = filePath.toLowerCase().replace(parentDir.toLowerCase(), '');

            if (this.tocFileRegex.test(input)) {
                fileInfoList.push(filePath);
            } else if (this.bindingsXmlRegex.test(input)) {
                matchingFileList.push(filePath);
            }
        }

        for (let fileInfo of fileInfoList) {
            await this.processIncludeFile(matchingFileList, fileInfo);
        }

        return matchingFileList;
    }

    private async processIncludeFile(matchingFileList: string[], fileInfo: string) {
        if (!fs.existsSync(fileInfo) || matchingFileList.indexOf(fileInfo) !== -1) {
            return;
        }

        matchingFileList.push(fileInfo);

        let input = await this._fileService.readFile(fileInfo);
        input = this.removeComments(fileInfo, input);

        const inclusions = this.getFileInclusionMatches(fileInfo, input);
        if (!inclusions || !inclusions.length) {
            return;
        }

        for (let include of inclusions) {
            const fileName = path.join(path.dirname(fileInfo), include);
            await this.processIncludeFile(matchingFileList, fileName);
        }
    }

    private getFileInclusionMatches(fileInfo: string, fileContent: string): string[] | null {
        const ext = path.extname(fileInfo);
        switch (ext) {
            case '.xml':
                return this.matchAll(fileContent, this.bindingsXmlIncludesRegex);
            case '.toc':
                return this.matchAll(fileContent, this.tocFileIncludesRegex);
            default:
                return null;
        }
    }

    private removeComments(fileInfo: string, fileContent: string): string {
        const ext = path.extname(fileInfo);
        switch (ext) {
            case '.xml':
                return fileContent.replace(this.bindingsXmlCommentsRegex, '');
            case '.toc':
                return fileContent.replace(this.tocFileCommentsRegex, '');
            default:
                return fileContent;
        }
    }

    private matchAll(str: string, regex: RegExp): string[] {
        const matches: string[] = [];
        let currentMatch: RegExpExecArray;
        do {
            currentMatch = regex.exec(str);
            if (currentMatch) {
                matches.push(currentMatch[1]);
            }
        } while (currentMatch);

        return matches;
    }

    private computeNormalizedFileHash(filePath: string) {
        return this.computeFileHash(filePath, true);
    }

    private computeFileHash(filePath: string, normalizeWhitespace: boolean) {
        return this.computeHash(filePath, 0, normalizeWhitespace);
    }

    private computeStringHash(str: string): Promise<number> {
        return new Promise((resolve, reject) => {
            const eventHandler = (_evt: any, arg: CurseHashFileResponse) => {
                if (arg.error) {
                    return reject(arg.error);
                }

                resolve(arg.fingerprint);
            };

            const request: CurseHashFileRequest = {
                targetString: str,
                targetStringEncoding: 'ascii',
                responseKey: uuidv4(),
                normalizeWhitespace: false,
                precomputedLength: 0
            };

            this._electronService.ipcRenderer.once(request.responseKey, eventHandler);
            this._electronService.ipcRenderer.send(CURSE_HASH_FILE_CHANNEL, request);
        });
    }

    private computeHash(
        filePath: string,
        precomputedLength: number = 0,
        normalizeWhitespace: boolean = false
    ): Promise<number> {
        return new Promise((resolve, reject) => {
            const eventHandler = (_evt: any, arg: CurseHashFileResponse) => {
                if (arg.error) {
                    return reject(arg.error);
                }

                resolve(arg.fingerprint);
            };

            const request: CurseHashFileRequest = {
                responseKey: uuidv4(),
                filePath,
                normalizeWhitespace,
                precomputedLength
            };

            this._electronService.ipcRenderer.once(request.responseKey, eventHandler);
            this._electronService.ipcRenderer.send(CURSE_HASH_FILE_CHANNEL, request);
        });
    }

}